arxiv/integration/meta.py

```diff
@@ -0,0 +1,146 @@
+"""Meta tools for integrations."""
+
+import inspect
+
+
+class MetaIntegration(type):
+    """
+    Metaclass for context-bound integrations.
+
+    The purpose of this metaclass is to simplify writing integrations that
+    need to attach to the application or request context.
+
+    A typical pattern  when using an integration's method is to first check for
+    an instance of the integration on the current context, and then call the
+    method on that instance. In practice, this means either performing the
+    context check  every time the method is used (yuck) or implementing
+    module-level functions that wrap the integration class instance methods,
+    and check the context when called (also yuck).
+
+    Here's an example of what life was like back then:
+
+
+    .. code-block:: python
+
+       class FileManagementService(object):
+           '''Encapsulates a connection with the file management service.'''
+
+           def __init__(self, *args, **kwargs) -> None:
+               ...
+
+           def get_upload_status(self, upload_id: int) -> Upload:
+               ...
+
+       def init_app(app: object = None) -> None:
+           '''Set default configuration params for an application instance.'''
+           config = get_application_config(app)
+           config.setdefault('FILE_MANAGER_ENDPOINT', 'https://arxiv.org/')
+           config.setdefault('FILE_MANAGER_VERIFY', True)
+
+       def get_session(app: object = None) -> FileManagementService:
+           '''Get a new session with the file management endpoint.'''
+           config = get_application_config(app)
+           endpoint = config.get('FILE_MANAGER_ENDPOINT', 'https://arxiv.org/')
+           verify_cert = config.get('FILE_MANAGER_VERIFY', True)
+           return FileManagementService(endpoint, verify_cert=verify_cert)
+
+       def current_session() -> FileManagementService:
+           '''Get/create :class:`.FileManagementService` for this context.'''
+           g = get_application_global()
+           if not g:
+               return get_session()
+           elif 'filemanager' not in g:
+               g.filemanager = get_session()   # type: ignore
+           return g.filemanager    # type: ignore
+
+       @wraps(FileManagementService.get_upload_status)
+       def get_upload_status(upload_id: int) -> Upload:
+           '''See :meth:`FileManagementService.get_upload_status`.'''
+           return current_session().get_upload_status(upload_id)
+
```

Comment:
It seems like this could be something relatively simple:
```
def setup_web_app():
...
app['FileManagementService'] = FileManagementServiceImplXYZ( config.get(xxx)...)
```
Then to use it
```
import current_app from Flask
...
current_app.FileManagementService.get_upload(id)
```
That would give an error like ```object has no attribute 'FileManagementService'``` if the app was misconfigured.

Reply:
See https://github.com/arXiv/arxiv-base/pull/107#discussion_r266385406